app/src/src_nonfreenet/org/breezyweather/sources/metoffice/MetOfficeService.kt

```diff
@@ -0,0 +1,114 @@
+
+package org.breezyweather.sources.metoffice
+
+import android.content.Context
+import android.graphics.Color
+import breezyweather.domain.location.model.Location
+import breezyweather.domain.weather.wrappers.WeatherWrapper
+import dagger.hilt.android.qualifiers.ApplicationContext
+import io.reactivex.rxjava3.core.Observable
+import org.breezyweather.BuildConfig
+import org.breezyweather.R
+import org.breezyweather.common.exceptions.ApiKeyMissingException
+import org.breezyweather.common.preference.EditTextPreference
+import org.breezyweather.common.preference.Preference
+import org.breezyweather.common.source.ConfigurableSource
+import org.breezyweather.common.source.HttpSource
+import org.breezyweather.common.source.MainWeatherSource
+import org.breezyweather.common.source.SecondaryWeatherSourceFeature
+import org.breezyweather.settings.SourceConfigStore
+import retrofit2.Retrofit
+import javax.inject.Inject
+import javax.inject.Named
+
+class MetOfficeService @Inject constructor(
+    @ApplicationContext context: Context,
+    @Named("JsonClient") client: Retrofit.Builder
+) : HttpSource(), MainWeatherSource, ConfigurableSource {
+
+    override val id = "metoffice"
+    override val name = "Met Office"
+    override val privacyPolicyUrl = "https://www.metoffice.gov.uk/policies/privacy"
+
+    override val color = Color.rgb(185, 213, 50)
+    override val weatherAttribution = "Met Office"
+
+    private val mApi by lazy {
+        client
+            .baseUrl(MET_OFFICE_BASE_URL)
+            .build()
+            .create(MetOfficeApi::class.java)
+    }
+
+    override val supportedFeaturesInMain = emptyList<SecondaryWeatherSourceFeature>()
+
+    override fun requestWeather(
+        context: Context, location: Location, ignoreFeatures: List<SecondaryWeatherSourceFeature>
+    ): Observable<WeatherWrapper> {
+        if (!isConfigured) {
+            return Observable.error(ApiKeyMissingException())
+        }
+
+        val apiKey = getApiKeyOrDefault()
+
+        return Observable.zip(
+            mApi.getHourlyForecast(apiKey, location.latitude, location.longitude),
+            mApi.getDailyForecast(apiKey, location.latitude, location.longitude)
+        ) { hourly, daily -> convert(hourly, daily) }
+    }
+
+
+    // CONFIG
+    private val config = SourceConfigStore(context, id)
+    private var apikey: String
+        set(value) {
+            config.edit().putString("apikey", value).apply()
+        }
+        get() = config.getString("apikey", null) ?: ""
+
+    private fun getApiKeyOrDefault(): String {
+        return apikey.ifEmpty { BuildConfig.MET_OFFICE_KEY }
+    }
+    override val isConfigured
+        get() = getApiKeyOrDefault().isNotEmpty()
+
+    override val isRestricted
+        get() = apikey.isEmpty()
+
+    override fun getPreferences(context: Context): List<Preference> {
+        return listOf(
+            EditTextPreference(
+                titleId = R.string.settings_weather_source_pirate_weather_api_key,
```

Comment:
You can't use Pirate Weather title ID. Although untested, I'm pretty sure this will crash the Weather sources settings screen

Reply:
Thanks, that was an oversight - I have fixed it now.